src/services/blockchain/ethereum/ethereum-blockchain-service.ts

```diff
@@ -269,91 +268,93 @@ export default class EthereumBlockchainService implements BlockchainService {
    * Sends transaction with root CID as data
    */
   public async sendTransaction(rootCid: CID): Promise<Transaction> {
-    const walletBalance = await this.wallet.provider.getBalance(this.wallet.address)
-    logMetric.ethereum({
-      type: 'walletBalance',
-      balance: ethers.utils.formatUnits(walletBalance, 'gwei'),
-    })
-    logger.imp(`Current wallet balance is ` + walletBalance)
-
-    const txData = await this._buildTransactionRequest(rootCid)
-    const transactionTimeoutSecs = this.config.blockchain.connectors.ethereum.transactionTimeoutSecs
-    const { network } = this.config.blockchain.connectors.ethereum
-
-    let attemptNum = 0
-    const txResponses: Array<providers.TransactionResponse> = []
-    while (attemptNum < MAX_RETRIES) {
-      try {
-        await this.setGasPrice(txData, attemptNum)
-
-        const txResponse = await this._trySendTransaction(txData, attemptNum, network)
-        txResponses.push(txResponse)
-        return await this._confirmTransactionSuccess(txResponse, network, transactionTimeoutSecs)
-      } catch (err) {
-        logger.err(err)
-
-        const { code } = err
-        if (code) {
-          if (code === ErrorCode.INSUFFICIENT_FUNDS) {
-            const txCost = (txData.gasLimit as BigNumber).mul(txData.gasPrice)
-            if (txCost.gt(walletBalance)) {
+    return this.withWalletBalance(async (walletBalance) => {
```

Comment:
ugh I see now.  You're trying to clean up how we log the final balance when the function ends. I really hate the extra scope, and the use of passing in a function to call that isn't actually a callback and will be executed inline is weird to me as well.  Feels like a whole lot of unnecessary machinery.  But I do agree that the random lines of code after the while loop that have to be remembered, and that prevents an early-return anywhere in main body of the function, isn't ideal either.  Sigh this would be so much cleaner in so many other languages.  RAII with C++ or Rust, or a `defer` statement in Go, would all be so much nicer for this.

Ultimately, while I personally would probably have left the structure closer to what it was, it's not really clear cut which approach is cleaner, so this seems fine to me if you like it better this way.